core/read.go

```diff
@@ -475,6 +475,144 @@ func readSymbol(reader *Reader, first rune) Object {
 	}
 }
 
+/* When validating symbol/keyword names, which is done only in
+   LINTER_MODE given the appropriate :rules in place, use function
+   variables for a) simplicity of functions, b) ease of adding new
+   ones (if new rules are desired), and c) hope for reasonably good
+   performance. */
+
+/* Returns whether a rune is a character that is inherently allowed in
+/* identifiers (symbols, keywords) by dint of the fact that
+/* clojure.core and other core packages define identifiers with these
+/* characters. While not important for parsing (Clojure is extremely
+/* permissive regarding which characters can be lexed into an
+/* identifier), linting can helpfully find and warn about characters
+/* outside of this set (as extended via configuration). */
+func isCoreIdentRune(r rune) bool {
+	switch r {
+	case '*', '+', '!', '-', '?', '=', '<', '>', '&', '_', '.', '\'': // Used in clojure.core, joker.core, etc.
+		return true
+	}
+	return ('a' <= r && r <= 'z') || ('A' <= r && r <= 'Z') || ('0' <= r && r <= '9')
+}
+
+const isValidCoreReason = "not a Letter nor (Decimal) Digit (category L nor Nd)"
+
+func isValidCore(r rune) bool {
+	return unicode.IsLetter(r) || unicode.IsDigit(r)
+}
+
+const isValidSymbolReason = "not a Letter, (Decimal) Digit, nor Symbol (category L, Nd, nor S)"
+
+func isValidSymbol(r rune) bool {
+	return unicode.IsLetter(r) || unicode.IsDigit(r) || unicode.IsSymbol(r)
+}
+
+const isValidVisibleReason = "not a Letter, (Decimal) Digit, Symbol, Punctuation, nor Mark (category L, Nd, S, P, nor M)"
+
+func isValidVisible(r rune) bool {
+	return unicode.IsLetter(r) || unicode.IsDigit(r) || unicode.IsSymbol(r) || unicode.IsPunct(r) || unicode.IsMark(r)
+}
+
+const isValidUnicodeReason = "not a Unicode character"
+
+func isValidUnicode(r rune) bool {
+	return r <= unicode.MaxRune
+}
+
+const isValidASCIIReason = "not a (7-bit) ASCII character"
+
+func isValidASCII(r rune) bool {
+	return r <= unicode.MaxASCII
+}
+
+const isValidAnyReason = "not anything!?"
+
+func isValidAny(r rune) bool {
+	return true
+}
+
+var identValidationSetFn = isValidCore
+var identValidationSetWhy = isValidCoreReason
+var identValidationRangeFn = isValidASCII
+var identValidationRangeWhy = isValidASCIIReason
+
+func warnInvalidIdent(reader *Reader, s *string) {
+	if s == nil {
+		return
+	}
+
+	for i, r := range *s {
+		if !isCoreIdentRune(r) && (!identValidationSetFn(r) || !identValidationRangeFn(r)) {
+			var explain string
+			if identValidationSetFn(r) {
+				explain = identValidationRangeWhy
+			} else if identValidationRangeFn(r) {
+				explain = identValidationSetWhy
+			} else {
+				explain = identValidationSetWhy + "; " + identValidationRangeWhy
+			}
+			runeValue, _ := utf8.DecodeRuneInString((*s)[i:])
```

Comment:
Would not `runeValue` be the same as `r`?

Reply:
> Would not `runeValue` be the same as `r`?

Ah, I'm sure it would now that the code works differently than the earlier version!

I'll do up a PR soon, after I take care of other things; or you can try it yourself.

(Which reminds me, I haven't gotten around to adding tests for this new linting capability yet; I just run my own hand-rolled ones....)

Reply:
No worries, I'll make the change and add tests.

Reply:
Made a few tweaks here: https://github.com/candid82/joker/commit/f3dd4c864978ef55330a43c31db279f1faa043f7